### DISCUSSION TITLE
Rule matching either some literals or specific words followed by numbers

### DISCUSSION THREAD

**Discussion by ero-qt** [antlr/antlr4#4677] [Q&A] [answered]
Hello,

sorry, I didn't know how else to explain it.

What I would like to make work is the following:

```g4
fragment LENGTH: [0-9]+;

STRING: 'string' LENGTH;
BYTE_ARRAY: 'byte' LENGTH;

type :
    'bool'
    | 'sbyte'
    | 'byte'
    | 'short'
    | 'ushort'
    | 'int'
    | 'uint'
    | 'long'
    | 'ulong'
    | 'float'
    | 'double'
    | STRING
    | BYTE_ARRAY
    ;
```

However, currently, I get an error saying `string32` is unexpected and it lists every member except for `string[0-9]+` and `byte[0-9]+`.  
How can I make this work?

Additionally, I seem to have a whitespace problem: `mismatched input '{\r\n}' expecting '{'`  
For this rule: `state : 'state' '(' STRING (',' STRING)? ')' '{' declaration+ '}'`

Why would this happen?

**Comment by kaby76** [comment]
There isn't enough information to form an answer. We need the entire grammar and input.

**Comment by ero-qt** [comment]
I had some things declared as parser rules when they should have been lexer rules. This was my whole problem.

Is there a document somewhere outline when something should be a a lexer rule or a parser rule? And when we should use fragments?

**Comment by kaby76** [comment, marked as answer]
* There are no hard rules for where to draw the boundary between lexer and parser. You could even define everything except a single character in the parser, but would likely suffer in performance.
* Study how it's done with other grammars. See [grammars-v4](https://github.com/antlr/grammars-v4).
* Use fragment when you want to create a reusable regular expression but where you don't want to generate a token.
* Make sure to understand the basic rules of how Antlr lexers work. See https://stackoverflow.com/questions/78856126/odata-parser-problem-with-substring-function#comment139034947_78856126
* Read StackOverflow.com answers for Antlr questions.
